python/ray/train/torch.py

```diff
@@ -195,27 +195,67 @@ class _WrappedDataLoader(DataLoader):
     def __init__(self, base_dataloader: DataLoader, device: torch.device):
 
         self.__dict__.update(getattr(base_dataloader, "__dict__", {}))
-        self.dataloader = base_dataloader
+        self._dataloader = base_dataloader
+        self.dataloader_iter = None
         self.device = device
+        # create a new CUDA stream to move data from host to device concurrently
+        self._memcpy_stream = torch.cuda.Stream() if device.type == "cuda" else None
+        self.curr_batch = None
+        self._prepare()
 
     def _move_to_device(self, item):
+        if item is None:
+            return None
+
         def try_move_device(i):
             try:
-                i = i.to(self.device)
+                i = i.to(self.device, non_blocking=True)
             except AttributeError:
                 logger.debug(f"Item {i} cannot be moved to device " f"{self.device}.")
             return i
 
-        return tuple(try_move_device(i) for i in item)
+        with torch.cuda.stream(self._memcpy_stream):
+            return tuple(try_move_device(i) for i in item)
+
+    def _wait_for_batch(self, item):
+        if self._memcpy_stream is None:
+            return
+        # Reference:
+        # https://pytorch.org/docs/stable/generated/torch.Tensor.record_stream.html
+        # The training stream (current) needs to wait until
+        # the memory copy stream finishes.
+        torch.cuda.current_stream().wait_stream(self._memcpy_stream)
+        curr_stream = torch.cuda.current_stream()
+        # When a tensor is used by CUDA streams different from
+        # its original allocator, we need to call ``record_stream``
+        # to inform the allocator of all these streams. Otherwise,
+        # the tensor might be freed once it is no longer used by
+        # the creator stream.
+        for i in item:
+            i.record_stream(curr_stream)
 
     def __len__(self):
-        return len(self.dataloader)
+        return len(self._dataloader)
 
-    def __iter__(self):
-        iterator = iter(self.dataloader)
+    def _prepare(self):
+        self.dataloader_iter = iter(self._dataloader)
+        self._get_next_batch()
 
-        for item in iterator:
-            yield self._move_to_device(item)
+    def _get_next_batch(self):
```

Comment:
nit: Rename this to something like `_prefetch_next_batch` since "get" usually implies that something will be returned.